game-core/src/main/java/games/strategy/triplea/delegate/battle/BattleState.java

```diff
@@ -30,10 +30,24 @@ public Side getOpposite() {
     }
   }
 
+  enum UnitsStatus {
+    // units that have not been hit
+    ALIVE,
+    // both ALIVE and CASUALTY
+    ACTIVE,
+    // units that have been hit but not yet cleared
+    CASUALTY,
+    // units that are dead
+    DEAD,
```

Comment:
Would it make sense to call this "removed" and use that terminology in the comments above as well?

I notice: " but not yet cleared", which would imply this is a "CLEARED" status. "CLEARED" could mean a few other things, "REMOVED" on the other hand I believe would be unambiguous. 

Arguably, 'dead' and 'casualty' can mean the same thing, a slight rename would keep that distinction crystal clear.

Reply:
Yeah, terminology is a little confusing.   Basically, there are 3 states a unit can have:

1. Unit has been removed from the battle
2. Unit is on the casualty strip but can still fire back
3. Unit is none of the above

The 2nd and 3rd case are generally requested together (such as when figuring out who can fire) so I added a 4th state to combine the 2nd and 3rd case.

So, the names I picked were:

1. DEAD
2. CASUALTY
3. ALIVE
4. ACTIVE (the union of CASUALTY and ALIVE)

Calling DEAD as REMOVED might add some confusion since retreated units are removed but they aren't dead.

I see the confusion where the term casualty generally includes both dead and wounded.  I thought of WOUNDED but a two-hit unit with only one hit is "wounded" but it wouldn't be in that group.  I choose CASUALTY since the rules generally call it the "casualty strip".

Reply:
Perhaps: `ACTIVE, CASUALTY, REMOVED_CASUALTY`

IMO it would be good to refer to RETREATED rather than REMOVED, but perhaps beyond the point.

The union of the two other states is kinda odd.  Is it actually an intersection? Can a unit have multiple states? To resolve this, perhaps the callers should query for either CASUALTY or ALIVE and then it's just on the caller to query for either state?

Reply:
I've updated the comments because the original comment said "ALIVE and CASUALTY" when it probably should have said "or".  It is a union of the two states, not an intersection.  A unit can only have one state.

The "ACTIVE" state is there to help ensure that code doesn't forget to check the CASUALTY state.  Right now, I think there are a few cases where it is forgetting.  I expect ACTIVE to be the normal state that callers will use and ALIVE and CASUALTY will be the special cases.

RETREATED doesn't exist yet but that is just because there is no current code that uses `BattleState` and that needs the retreated units.  It probably will exist at some point.

Is `REMOVED_CASUALTY` supposed to replace `DEAD`?  Because the `DEAD` state really are dead units.  They are no longer in the game.

Reply:
> Is REMOVED_CASUALTY supposed to replace DEAD? Because the DEAD state really are dead units. They are no longer in the game.

Yes, I believe the two are quite synonymous, a removed casualty is removed from the game. The idea is to try and bake into the names a relationship so we can see the transition from the naming alone. 'casualty' -> 'removed-casualty'. 

-----

The union still strikes me as odd. Typically states and even enum values are well to be mutually exclusive. It's kinda confusing as the union is both states, so it could be confusing to think that you are one enum value when you are really another. For example, take "A", "B", and "C" is the union of "A" and "B". If someone is looking at code dealing with only "A" and "B", it could be confusing to get a "C" state back and or know that if a unit transitions to state "B" then it's a conditional transition and the unit might have to go to state "C".

Any thoughts to having callers to send the union? We have a vararg method already, I think one would only need to replace "ALIVE" to be its constituent components. That would make it clear at a caller level we are doing an 'or' between two states.

Reply:
I thought of doing a method but java doesn't allow two varargs in the same method.  I've just pushed a change where I created an overloaded getUnits.  And I've removed ACTIVE.

But, maybe the confusion is that it is named "states".  You initially mentioned calling it `filterUnits`.  Maybe it should be changed to that.  What about `UnitBattleFilter` or something?  Because ACTIVE isn't a "state".  It is a filter that is the boolean OR of ALIVE and CASUALTY.  A unit is either ALIVE or CASUALTY.  And since I expect most callers to want both ALIVE and CASUALTY units, I wanted to make it simple for callers to just request a pre-built "OR" filter, called ACTIVE.

Reply:
Here's another idea.  What about keeping the `UnitBattleStatus` and add a `UnitBattleFilter` that is defined like:

```
  enum UnitBattleFilter {
    ACTIVE(EnumSet.of(UnitBattleStatus.ALIVE, UnitBattleStatus.CASUALTY)),
    ALIVE(EnumSet.of(UnitBattleStatus.ALIVE)),
    CASUALTY(EnumSet.of(UnitBattleStatus.CASUALTY));

    EnumSet<UnitBattleStatus> status;
    UnitBattleFilter(final EnumSet<UnitBattleStatus> status) {
      this.status = status;
    }
  }
```

Then `getUnits` (or `filterUnits`) would take a `UnitBattleFilter`.

Reply:
Interesting idea. My only concern is the amount of scaffolding/indirection, but we're still in a very reasonable realm here.

You might be able to absorb the filter method into the `UnitBattleFilter` method as well.

Some really rough code for that to give an idea:
```
public class UnitBattleFilter implements UnaryFunction<Unit> {
    ACTIVE = new UnitBattleFilter(UnitBattleStatus.ALIVE, UnitBattleStatus.CASUALTY));
   
   private final List<UnitBattleStatus> states;

    private    UnitBattleFilter(UnitBattleStatus ... states)  {
  }
```

Reply:
I don't think I can absorb the filter method into the `UnitBattleFilter`.  The class will need to access the raw unit lists (`attackingUnits`, `defendingUnits`, etc.) from `MustFightBattle` so I'd have to make public methods for access to those lists and that takes us back to having a bunch of public methods.

Reply:
Oh, I guess if the unit knew which status it was, then the unary function would work.  But unfortunately, the units don't actually know what their status is.  Only `MustFightBattle` does at this time.

Reply:
I've been wondering if unit attachment should be collapsed into units. That's a big change. The filter enum seems like a good compromise though in the meantime.

Reply:
That's a big change.  But even unit attachment doesn't store the battle status.  There's nothing on the unit nor anything attached to it (type, attachment, etc) that has the battle status.  It is solely managed by internal lists in `MustFightBattle`.  Which can mean that you can cause a bug where a unit is both alive, casualty, and dead at the same time.  You just have to add the unit to all three lists :)

That's kind of what caused the paratrooper issue.  The paratrooper was dead and so it was in the killed list.  But it was a dependent, so it was in the dependent list as well.  And air dependents are always considered alive.  So the paratrooper was both dead and alive :)